### DISCUSSION TITLE
TLS1.2 the curve when selecting the signature algorithm does not need to match exactly with the curve in the certificate public key

### DISCUSSION THREAD

**Discussion by lurio-liu** [openssl/openssl#23700] [Q&A] [answered]
In tls_choose_sigalg(), When the protocol is tls1.2, the signature algorithm curve does not need to be completely consistent with the public key curve in the certificate to pass, that is, the certificate of p256 can also choose the signature algorithm of p384. However, in tls1.3, it is necessary for the two to match exactly

**Comment by t8m** [comment]
Can you be please more specific - what is the question here?

**Reply by lurio-liu** [reply]
int tls_choose_sigalg(SSL *s, int fatalerrs){

        ......

        if (SSL_IS_TLS13(s) || SSL_IS_DTLS13(s)) {

         ......

         }else{

                ......

                int curve = -1;

                SSL_CTX *sctx = SSL_CONNECTION_GET_CTX(s);

                /* For Suite B need to match signature algorithm to curve */
                if (tls1_suiteb(s))
                    curve = ssl_get_EC_curve_nid(s->cert->pkeys[SSL_PKEY_ECC]
                                                 .privatekey);

                /*
                 * Find highest preference signature algorithm matching
                 * cert type
                 */
                for (i = 0; i < s->shared_sigalgslen; i++) {
                    lu = s->shared_sigalgs[i];

                    if (s->server) {
                        if ((sig_idx = tls12_get_cert_sigalg_idx(s, lu)) == -1)
                            continue;
                    } else {
                        int cc_idx = s->cert->key - s->cert->pkeys;

                        sig_idx = lu->sig_idx;
                        if (cc_idx != sig_idx)
                            continue;
                    }
                    /* Check that we have a cert, and sig_algs_cert */
                    if (!has_usable_cert(s, lu, sig_idx))
                        continue;
                    if (lu->sig == EVP_PKEY_RSA_PSS) {
                        /* validate that key is large enough for the signature algorithm */
                        EVP_PKEY *pkey = s->cert->pkeys[sig_idx].privatekey;

                        if (!rsa_pss_check_min_key_size(sctx, pkey, lu))
                            continue;
                    }
                    if (curve == -1 || lu->curve == curve)
                        break;
                }
               ......

        }
    }

in TLS1.2:
In this function, if we choose the ECC curve, not Suite B, and the variable curve=-1.
when we select an ECC certificate, we do not determine whether the public key curve in the certificate matches the curve in the signature algorithm in the following judgment.
if (curve == -1 || lu->curve == curve)，By using this judgment, we directly break.

`

    static const SIGALG_LOOKUP *find_sig_alg(SSL_CONNECTION *s, X509 *x,EVP_PKEY *pkey){
    
    const SIGALG_LOOKUP *lu = NULL;

    size_t i;

    int curve = -1;

    EVP_PKEY *tmppkey;

    SSL_CTX *sctx = SSL_CONNECTION_GET_CTX(s);

    /* Look for a shared sigalgs matching possible certificates */
    for (i = 0; i < s->shared_sigalgslen; i++) {
        lu = s->shared_sigalgs[i];

        /* Skip SHA1, SHA224, DSA and RSA if not PSS */
        if (lu->hash == NID_sha1
            || lu->hash == NID_sha224
            || lu->sig == EVP_PKEY_DSA
            || lu->sig == EVP_PKEY_RSA)
            continue;
        /* Check that we have a cert, and signature_algorithms_cert */
        if (!tls1_lookup_md(sctx, lu, NULL))
            continue;
        if ((pkey == NULL && !has_usable_cert(s, lu, -1))
                || (pkey != NULL && !is_cert_usable(s, lu, x, pkey)))
            continue;

        tmppkey = (pkey != NULL) ? pkey
                                 : s->cert->pkeys[lu->sig_idx].privatekey;

        if (lu->sig == EVP_PKEY_EC) {
            if (curve == -1)
                curve = ssl_get_EC_curve_nid(tmppkey);
            if (lu->curve != NID_undef && curve != lu->curve)
                continue;
        } else if (lu->sig == EVP_PKEY_RSA_PSS) {
            /* validate that key is large enough for the signature algorithm */
            if (!rsa_pss_check_min_key_size(sctx, tmppkey, lu))
                continue;
        }
        break;
    }

    if (i == s->shared_sigalgslen)
        return NULL;

    return lu;
    }

`
in TLS1.3:
in this function,When the signature algorithm is EVP_PKEY-EC, it will check whether the curve in the certificate matches the curve of this signature algorithm。

My question is, why doesn't the curve of the certificate in tls1.2 need to be exactly the same as the curve in the signature algorithm when the signature algorithm is of type EVP_PKEY-EC?

**Reply by mattcaswell** [reply, marked as answer]
The interpretation of sigalgs values changed between TLSv1.2 and TLSv1.3.

In TLSv1.3 the contents of the "signature_algorithms" extension is interpreted as a value from the IANA SignatureScheme list:

https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml#tls-signaturescheme

An example element in this list might be "ecdsa_secp384r1_sha384". This specifies the signature algorithm itself, the digest, and the curve that must be used.

In TLSv1.2 the contents of the "signature_algorithms" extension is a concatenation of the SignatureAlgorithms list and the HashAlgorithms list:

https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml#tls-parameters-16
https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml#tls-parameters-18

An example SignatureAlgorithm would be "ecdsa", while a hash algorithm would be "sha384". Not *no curve* is specified here.

So the value 0x0503 means "ecdsa_secp384r1_sha384" in TLSv1.3. In TLSv1.2 it means "ecdsa+sha384". So the standard does not imply any curve in TLSv1.2 - but in TLSv1.3 it does.

**Reply by lurio-liu** [reply]
Understood, thank you.
